refactor the exit of nvda and gui.terminate

seanbudd opened the pull request:
### Link to issue number:

Closes #12238

### Summary of the issue:

When restarting NVDA, `WM_QUIT` is posted as an event to the window, forcibly exiting the app. This leaves objects such as the system tray icon left behind. 

Additionally, changes introduced in #12183

- caused the braille viewer to be closed without saving state properly
- lost code that destroyed the system tray and menu in some instances
 - made most of `gui.terminate` no longer necessary/redundant

### Description of how this pull request fixes the issue:

- A windows event `winUser.WM_EXIT_NVDA` is registered that triggers `safeAppExit` and can be called across instances of NVDA. 
- move the safe destruction of the brailleviewer to `safeAppExit` so that it is exited properly before destruction
- reintroduce the destruction of the system tray icon and menu, and remove the icon manually.
- ensured `safeAppExit` is not called from `gui.terminate` if it has been called elsewhere to terminate the app. `WM_QUIT` is the other way to exit the MainLoop other than `safeAppExit`
- removed restarting the MainLoop in `gui.terminate` to process pending events as this doesn't work. 

### Testing strategy:

- Ensure NVDA system tray icons are cleaned on any exit. 
- Ensure NVDA exits properly on restart and when starting the process subsequently. Check `nvda-old.log` when restarting NVDA. 
- Ensure NVDA can exit and restart correctly without errors or crashes. Existing system tests exist for this process. 
- Running 2020.4 (leave it running) and start this build. This build should cause 2020.4 to exit, and this build should replace it as the running version. Confirm by checking about -> version
- Running this build (leave it running) and start 2020.4 which should cause this build to exit, and 2020.4 should replace it as the running version. Confirm by checking about -> version

### Known issues with pull request:

When starting a new instance of NVDA with an existing instance running, where one is version `<2020.4`, NVDA will not exit safely. Instead, the running NVDA copy will terminate directly using the behaviour of `2020.4`. This is because `WM_EXIT_NVDA` won't be registered on the older instance.  

Issues with terminating NVDA across instances cannot be logged properly as the loghandler hasn't been initialized

### Change log entry:

Developer changes

```markdown
-  do not exit NVDA by sending a `WM_QUIT` message to the process. Instead send `winUser.WM_EXIT_NVDA` to the window handle (found by `winUser.FindWindow('wxWindowClassNR', 'NVDA')`)
```

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

seanbudd committed:
refactor the exit of nvda

seanbudd committed:
fix lint

seanbudd committed:
remove systray icon before destroying

seanbudd committed:
safely exit NVDA across instances

seanbudd committed:
fix log strings

feerrenrut commented:
Can you also test:
- Running 2020.4 (leave it running) and start this build. This build should cause 2020.4 to exit, and this build should replace it as the running version. Confirm by checking about -> version
- Running this build (leave it running) and start 2020.4 which should cause this build to exit, and 2020.4 should replace it as the running version. Confirm by checking about -> version

seanbudd commented:
@feerrenrut - these test cases work (as in the correct version of NVDA runs) but NVDA won't exit safely, I've noted this in the `issues with this PR` section.

seanbudd committed:
revert removal of register of WM_QUIT

seanbudd commented:
will this backwards upgrade compatibility need to stay forever? I've marked it for 2022.1 deprecation but it would still be an issue when upgrading from 2020.1.

feerrenrut commented:
> but NVDA won't exit safely, I've noted this in the issues with this PR section.

That's fine. I don't think there is any way around that.

> will this backwards compatibility for upgrading/downgrading need to stay forever?

I think it should stay for a long time. I wouldn't remove it for 2022.1.
Going forward, newer version (eg 2021.1 and later) won't have the limitation from the "known issues" section, right? If so, please clarify.

seanbudd commented:
> I think it should stay for a long time. I wouldn't remove it for 2022.1.

How should I mark the code? Or should the deprecation notes be removed entirely?

> Going forward, newer version (eg 2021.1 and later) won't have the limitation from the "known issues" section, right? If so, please clarify.

Correct, how does the new message look?

feerrenrut commented on source/gui/__init__.py:
```diff
@@ -582,27 +596,7 @@ def wx_CallAfter_wrapper(func, *args, **kwargs):
-	import brailleViewer
-	brailleViewer.destroyBrailleViewer()
-
-	for instance, state in gui.SettingsDialog._instances.items():
-		if state is gui.SettingsDialog._DIALOG_DESTROYED_STATE:
-			log.error(
-				"Destroyed but not deleted instance of settings dialog exists: {!r}".format(instance)
-			)
-		else:
-			log.debug("Exiting NVDA with an open settings dialog: {!r}".format(instance))
```
Please keep this code. Happy for it to be moved to another function with a comment from:
https://github.com/nvaccess/nvda/pull/12291#pullrequestreview-636204844

seanbudd replied:
I've kept and moved the changed code from #12291 to `safeAppExit`.

feerrenrut commented on source/nvda.pyw:
```diff
@@ -251,7 +256,9 @@ if customVenvDetected:
+# Accept wm_quit from other processes, even if running with higher privilages
+# This allows for downgrading between NVDA versions, deprecated in 2022.1
+# in 2022.1 remove the call to ChangeWindowMessageFilter (#12286)
 if not ctypes.windll.user32.ChangeWindowMessageFilter(winUser.WM_QUIT,1):
```
Can you make `1` a constant or enum: 
MSGFLT_ADD=1 # Adds the message to the filter. This has the effect of allowing the message to be received.
MSGFLT_REMOVE=2 # Removes the message from the filter. This has the effect of blocking the message.

seanbudd replied:
added these to `winUser.py`

feerrenrut commented on source/nvda.pyw:
```diff
@@ -251,7 +256,9 @@ if customVenvDetected:
 if globalVars.appArgs.changeScreenReaderFlag:
 	winUser.setSystemScreenReaderFlag(True)
-#Accept wm_quit from other processes, even if running with higher privilages
+# Accept wm_quit from other processes, even if running with higher privilages
+# This allows for downgrading between NVDA versions, deprecated in 2022.1
+# in 2022.1 remove the call to ChangeWindowMessageFilter (#12286)
```
I'd say some thing like "2020.4 and earlier versions sent a WM_QUIT message when asking NVDA to exit. Some users may run several different versions of NVDA, so we continue to support this. "

I don't think we need to mark it as deprecated, or have a plan for removal at this stage. I wouldn't consider removing this for many years.

If it is still true, you might also add "WM_QUIT does not allow NVDA to shutdown cleanly, now WM_EXIT_NVDA is used instead".

feerrenrut commented on source/winUser.py:
```diff
@@ -377,6 +377,11 @@ class GUITHREADINFO(Structure):
+# Registers an application wide Window Message so that NVDA can be exited across instances
+WM_EXIT_NVDA = user32.RegisterWindowMessageW("WM_EXIT_NVDA")
+if not WM_EXIT_NVDA:
+	raise WinError()
```
Does this result in a log message that is easy to understand?

seanbudd replied:
I've added some better logging and comments across all the errors and error handling

seanbudd committed:
Merge branch 'master' into cleanup-exit

seanbudd committed:
incorporate review feedback

seanbudd commented on source/core.py:
```diff
@@ -391,13 +391,20 @@ def __init__(self, windowName=None):
 				self.handlePowerStatusChange()
 			elif msg == winUser.WM_DISPLAYCHANGE:
 				self.handleScreenOrientationChange(lParam)
+			elif msg == winUser.WM_EXIT_NVDA:
```
Unfortunately no, I tested this to be sure but [the API docs](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) says

>The WM_QUIT message is not associated with a window and therefore will never be received through a window's window procedure. It is retrieved only by the GetMessage or PeekMessage functions.

michaelDCurran commented:
Is quitting a newer copy of NVDA by version <= 2020.4 (with WM_QUIT) any worse now than it used to be? Granted it probably never entirely worked, but has the movement of code now caused less things to be cleaned up? 
It should be said though that I do agree with the changes and like the fact we have moved away from WM_QUIT.

feerrenrut commented on source/winUser.py:
```diff
@@ -377,10 +377,23 @@ class GUITHREADINFO(Structure):
 # The height of the virtual screen, in pixels.
 SM_CYVIRTUALSCREEN = 79
 
+# Actions associated with ChangeWindowMessageFilterEx
+# https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-changewindowmessagefilterex
+# Adds the message to the filter. This has the effect of allowing the message to be received.
+MSGFLT_ALLOW = 1
+# Removes the message from the filter. This has the effect of blocking the message.
+MSGFLT_DISALLOW = 2
+# Resets the window message filter to the default.
+# Any message allowed globally or process-wide will get through.
+MSGFLT_RESET = 0
```
I think over time we should try to move towards enums for these constants. The downside is that the names no longer exactly match the constant names in Windows. 

The upside is that it helps to make it clear what are the accepted values for the argument, it is harder to use a constant intended for a different function. The other benefit is that it becomes easier to trace where in the codebase these values are being used.

```suggestion
class MSGFLT(enum.IntEnum):
	# Actions associated with ChangeWindowMessageFilterEx
	# https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-changewindowmessagefilterex
	# Adds the message to the filter. This has the effect of allowing the message to be received.
	ALLOW = 1
	# Removes the message from the filter. This has the effect of blocking the message.
	DISALLOW = 2
	# Resets the window message filter to the default.
	# Any message allowed globally or process-wide will get through.
	RESET = 0

```

I don't think this is super important, I'll leave this up to you.

seanbudd commented:
> Is quitting a newer copy of NVDA by version <= 2020.4 (with WM_QUIT) any worse now than it used to be? Granted it probably never entirely worked, but has the movement of code now caused less things to be cleaned up?
> It should be said though that I do agree with the changes and like the fact we have moved away from WM_QUIT.

No it should be no different, `WM_QUIT` kills the process instantly and doesn't trigger any clean up code.

seanbudd committed:
Use enum for constants

Co-authored-by: Reef Turner <feerrenrut@users.noreply.github.com>

seanbudd committed:
Merge branch 'master' into cleanup-exit

seanbudd committed:
Merge branch 'cleanup-exit' of https://github.com/nvaccess/nvda into cleanup-exit

seanbudd committed:
use enum

seanbudd committed:
call safeAppExit if it is not called in terminate

seanbudd commented:
> Is quitting a newer copy of NVDA by version <= 2020.4 (with WM_QUIT) any worse now than it used to be? Granted it probably never entirely worked, but has the movement of code now caused less things to be cleaned up?
> It should be said though that I do agree with the changes and like the fact we have moved away from WM_QUIT.

No, as `WM_QUIT` causes the `MainLoop` to exit without calling `safeAppExit`. I've added some code to call `safeAppExit` in `gui.terminate` if it hasn't been called yet https://github.com/nvaccess/nvda/pull/12286/commits/9f561ca33dc5bf5ffe8be1a4577ff5326560157e .

michaelDCurran commented:
In safeAppExit: I think we should replace all x.Destroy() and wx.CallAfter(x.Destroy) calls with app.ScheduleForDestruction(x).
app being gotten from wx.GetApp()
ScheduleForDestruction, given any wx object (so I'm assuming windows, menus, icons etc), will schedule that object for destruction in the next event loop cycle, or if not in the event loop (UsesEventLoop() is False such as after exiting the MainLoop) then ScheduleForDestruction will destroy the objects in-place.
So, this should handle SafeAppExit being called either inside the event loop such as for a clean exit with wm_exit_nvda, or outside of the event loop, such as with wm_quit.
I have not tested this though, just going by the docs and WXWidgets source code. But I specifically looked into this as we have calls to wx.CallAfter in safeAppExit, but it may be called outside of the event loop in gui.terminate.

seanbudd committed:
Merge branch 'master' into cleanup-exit

seanbudd committed:
remove duplicated WM exit codes

seanbudd committed:
use ScheduleForDestruction

seanbudd committed:
ensure WM_QUIT is sent on WM_EXIT_NVDA failure

seanbudd commented:
@michaelDCurran  - I've incorporated that change and retested. I noticed that the WM_QUIT signal was not being sent properly (based off my misunderstanding of how WM_QUIT worked from earlier). This should be fixed now.

seanbudd committed:
update changes

michaelDCurran commented:
The print calls in nvda.pyw (which happen when upgrading from an older snapshot build, or any older release) are causing a wx error dialog to appear in binary builds. This is confusing, but is also seeming to freeze up the Windows shell, or at very least, NVDA is struggling to read much after this occurs. 
 Purely for the error dialog with the print calls alone, I think we must revert this pr for now and investigate further.

lukaszgo1 commented:
I'm getting a lot of crashes when restarting binary  versions of NVDA after this PR was merged with either [WinError 1400] Invalid window handle or [WinError 5] Access denied

seanbudd commented:
1. Revert PR
1. Reopen this PR and:
   - Add add-on handler clean up to `safeAppExit`
   - Remove print statements
   - Change the exit process by renaming window with the version number, check if that is there to decide whether to send `WM_QUIT` or `WM_EXIT_NVDA`
   - test the binaries
